source/common/http/hash_policy.cc

```diff
@@ -39,14 +42,39 @@ class HeaderHashMethod : public HashMethodImplBase {
                                     const StreamInfo::FilterStateSharedPtr) const override {
     absl::optional<uint64_t> hash;
 
-    // TODO(mattklein123): Potentially hash on all headers.
     const auto header = headers.get(header_name_);
     if (!header.empty()) {
+      const bool check_multiple_values =
+          Runtime::runtimeFeatureEnabled("envoy.reloadable_features.hash_multiple_header_values");
+      absl::InlinedVector<std::string, 1> rewritten_header_values;
+      absl::InlinedVector<absl::string_view, 1> header_values;
+
+      size_t num_headers_to_hash = 1;
+      if (check_multiple_values) {
+        num_headers_to_hash = header.size();
+        header_values.reserve(num_headers_to_hash);
+      }
+
+      for (size_t i = 0; i < num_headers_to_hash; i++) {
+        header_values.push_back(header[i]->value().getStringView());
+      }
+
       if (regex_rewrite_ != nullptr) {
-        hash = HashUtil::xxHash64(regex_rewrite_->replaceAll(header[0]->value().getStringView(),
-                                                             regex_rewrite_substitution_));
+        rewritten_header_values.reserve(num_headers_to_hash);
+        for (auto& value : header_values) {
+          rewritten_header_values.push_back(
+              regex_rewrite_->replaceAll(value, regex_rewrite_substitution_));
+          value = rewritten_header_values.back();
+        }
+      }
+
+      if (check_multiple_values && (num_headers_to_hash > 1)) {
```

Comment:
I think Matt commented above but you don't this conditional. num_headers_hash can be set to a value greater than one online if check_multiple_values is true, and the sorting will be a no-op for size==1.

Reply:
that is what Matt is look for, better to have @mattklein123 check again. also I have comment about the runtime flag above, it may needn't this check anymore.

Reply:
See my above comment. I think we do need the flag, but we can just set the size to 1 or not above and then we can collapse all of this logic.

Reply:
I updated, hope I understand correctly.